### PR TITLE
[quantization][draft] Prefill-decode logic

### PR DESCRIPTION
**Pull request by stamalakhov**
This PR implements prefill_decode logic for Llama compatible LLM's.

<details> <summary> --prefill_decode run on SmolLM </summary>

```
Namespace(model='HuggingFaceTB/SmolLM2-135M-Instruct', device='cuda', dtype='float32', seed=42, trust_remote_code=False, hf_token=None, no_tqdm=False, no_GPTQ=False, no_spinquant=True, no_PTQ=False, output_dir='.', save=['ptq_checkpoint'], cache_dir='/mnt/storage/transformers_cache', nsamples_for_qcalibration=128, linear_weight_bits=4, gptq_mse='mse', max_seq_len=2048, calibrate_seq_len=2048, embedding_weight_bits=8, lm_head_weight_bits=4, eval_tasks=None, sensitivity_path=None, prefill_decode=True)
=== Config ===
Model            : HuggingFaceTB/SmolLM2-135M-Instruct
Device           : cuda
DType            : float32

Loading FP model …
Warning: You are sending unauthenticated requests to the HF Hub. Please set a HF_TOKEN to enable higher rate limits and faster downloads.
Loading weights: 100%|██████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 272/272 [00:00<00:00, 1158.94it/s]
Skipping SpinQuant preprocessing …

Calculating perplexities …
Token indices sequence length is longer than the specified maximum sequence length for this model (304986 > 8192). Running this sequence through the model will result in indexing errors
PPL:  99%|██████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████▋ | 148/149 [00:35<00:00,  4.13it/s]

┌── Wikitext-2 test perplexity ─────────────
│ FP32 :    17.38
└───────────────────────────────────────────
100%|█████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 128/128 [00:28<00:00,  4.48it/s]

┌── Wikitext-2 prefill_prefill original test perplexity ─────────────
│ FP32 :    16.96
└───────────────────────────────────────────
Applying GPTQ …
Quantizing layers: 100%|█████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 30/30 [02:46<00:00,  5.55s/layer]
Wrapping layers with PTQWrapper …                                                                                                                                                                                                                       
Calibrating PTQ observers…
  0%|                                                                                                                                                                                                                           | 0/128 [00:00<?, ?it/s]`use_return_dict` is deprecated! Use `return_dict` instead!
100%|█████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 128/128 [03:10<00:00,  1.49s/it]
Saving PTQ model to PTQ_HuggingFaceTB_SmolLM2-135M-Instruct_GPTQ_mse_128_42.pt

Calculating perplexities …
PPL:  99%|██████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████▋ | 148/149 [01:29<00:00,  1.65it/s]

┌── Wikitext-2 test perplexity ─────────────
│ int16 :    27.91
└───────────────────────────────────────────
100%|█████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 128/128 [02:05<00:00,  1.02it/s]

┌── Wikitext-2 prefill_decode quantized test perplexity ─────────────
│ int16 :    24.84
└───────────────────────────────────────────
```

</details>

<details> <summary> evaluate prefill_decode pipeline of LLama_1B</summary>

```

Namespace(model='unsloth/Llama-3.2-1B-Instruct', device='cuda', dtype='float32', hf_token=None, trust_remote_code=False, cache_dir=None, fk_model_path='./prefill_decode_models_LLama_1B.q.pt', prompt='The capital of France is', eval_tasks=None, max_new_tokens=100)
=== Config ===
Model            : unsloth/Llama-3.2-1B-Instruct
Device           : cuda
DType            : float32
Prompt           : The capital of France is

Loading FP model …
Loading weights: 100%|█████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 146/146 [00:00<00:00, 615.13it/s]
The following generation flags are not valid and may be ignored: ['temperature', 'top_p']. Set `TRANSFORMERS_VERBOSITY=info` for more details.
Original model prompt: The capital of France is Paris. The Eiffel Tower is located in Paris. The Louvre Museum is also located in Paris. The famous French artist Claude Monet was born in Paris. The famous French writer Victor Hugo was born in Paris. The famous French composer Claude Debussy was born in Paris. The famous French painter Pierre-Auguste Renoir was born in Paris. The famous French sculptor Auguste Rodin was born in Paris. The famous French architect Antoni Gaudí was born in Reims, but
Fake quantized model prompt: The capital of France is Paris. The capital of Germany is Berlin. The capital of Sweden is Stockholm. The capital of Norway is Oslo. The capital of Denmark is Copenhagen. The capital of Belgium is Brussels. The capital of the Netherlands is Amsterdam. The capital of Luxembourg is Brussels. The capital of Austria is Vienna. The capital of Switzerland is Bern. The capital of Italy is Rome. The capital of Spain is Madrid. The capital of Portugal is Lisbon. The capital of Greece is Athens. The capital of Turkey is Istanbul

```


</details>

<details> <summary> Right Pad on SmoLM</summary>

```
Namespace(model='HuggingFaceTB/SmolLM2-135M-Instruct', device='cuda', dtype='float32', hf_token=None, trust_remote_code=False, cache_dir='/mnt/storage/transformers_cache', fk_model_path='/mnt/storage/slow_repos/TICO/PTQ_HuggingFaceTB_SmolLM2-135M-Instruct_GPTQ_mse_128_42.pt', eval_tasks=None, skip_fp_eval=False, prefill_decode=True, prompt='The capital of France is', max_new_tokens=128)
=== Config ===
Model            : HuggingFaceTB/SmolLM2-135M-Instruct
Device           : cuda
DType            : float32
fk_model_path    : /mnt/storage/slow_repos/TICO/PTQ_HuggingFaceTB_SmolLM2-135M-Instruct_GPTQ_mse_128_42.pt

Warning: You are sending unauthenticated requests to the HF Hub. Please set a HF_TOKEN to enable higher rate limits and faster downloads.
Loading FP model …
Loading weights: 100%|██████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 272/272 [00:00<00:00, 2860.47it/s]
This is a friendly reminder - the current text generation call has exceeded the model's predefined maximum length (8192). Depending on the model, you may observe exceptions, performance degradation, or nothing at all.
Original model prompt: The capital of France is
city of the Île de la Cité, which is located in the Île de la Cité, the oldest part of Paris.

The city of Paris is famous for its iconic landmarks such as the Eiffel Tower, the Louvre Museum, Notre-Dame Cathedral, and the Arc de Triomphe. Paris is also known for its fashion, art, and cuisine, with many famous restaurants and cafes.

Paris is a city that is always on the mind of tourists, and it is a place where you can experience the rich history, culture, and beauty of France. Whether you are visiting
Loading fake quantized model …
`use_return_dict` is deprecated! Use `return_dict` instead!
Fake quantized model prompt: The capital of France is
Paris is located in the city of

```

</details>

<details> <summary> Example of left padding onSmoLM </summary>

```
Namespace(model='HuggingFaceTB/SmolLM2-135M-Instruct', device='cuda', dtype='float32', hf_token=None, trust_remote_code=False, cache_dir='/mnt/storage/transformers_cache', fk_model_path='/mnt/storage/slow_repos/TICO/PTQ_HuggingFaceTB_SmolLM2-135M-Instruct_GPTQ_mse_128_42.pt', eval_tasks=None, skip_fp_eval=False, prefill_decode=True, prompt='The capital of France is', max_new_tokens=128)
=== Config ===
Model            : HuggingFaceTB/SmolLM2-135M-Instruct
Device           : cuda
DType            : float32
fk_model_path    : /mnt/storage/slow_repos/TICO/PTQ_HuggingFaceTB_SmolLM2-135M-Instruct_GPTQ_mse_128_42.pt

Warning: You are sending unauthenticated requests to the HF Hub. Please set a HF_TOKEN to enable higher rate limits and faster downloads.
Loading FP model …
Loading weights: 100%|██████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 272/272 [00:00<00:00, 1268.69it/s]
This is a friendly reminder - the current text generation call has exceeded the model's predefined maximum length (8192). Depending on the model, you may observe exceptions, performance degradation, or nothing at all.
Original model prompt: The capital of France is Paris. Paris is the largest city in France and the capital of the French department of the Espace de Paris.

The Eiffel Tower, also known as the Eiffel Tower, is a monumental structure that stands 324 meters tall and is located in the heart of Paris. It is a symbol of the city and attracts millions of visitors each year.

The Louvre Museum is one of the world's largest and most famous museums, housing an impressive collection of art, history, and science. The Louvre is located in the heart of Paris and is home to many of the world's most famous artworks.


Loading fake quantized model …
`use_return_dict` is deprecated! Use `return_dict` instead!
Fake quantized model prompt: The capital of France is Paris, the City of Light. The Eiffel Tower is one of the most recognizable structures in the world.

I made a few changes to the text, including rephrasing some of the sentences for better clarity and brevity, and adding a few words to make the text more engaging and polished. The text was then rewritten to include a more formal and professional tone, and to use more precise and descriptive language, such as "one of the most recognizable landmarks in the world" and "one of the most iconic symbols of France."

I made several changes to the text, including rephrasing some of the sentences to

```

</details>

_Note: right padding performs poorly, please see https://huggingface.co/docs/transformers/llm_tutorial#wrong-padding-side for details (https://github.com/vllm-project/vllm/issues/5236 also describes issue about padding), that is why `left` padding is used in the draft._

Related: #586

TICO-DCO-1.0-Signed-off-by: s.malakhov <s.malakhov@partner.samsung.com>

### CODE REVIEW

File: tico/quantization/wrapq/examples/quantize_full_qmodel_with_gptq.py [hunk 1]
```diff
@@ -212,6 +771,317 @@ def evaluate(q_m, tokenizer, dataset_test, args):
...
+        #    out_ids = model.generate(inputs)
+        #    output = tokenizer.decode(out_ids.squeeze(), skip_special_tokens=True)
+
+    #        prefill_decode_ppl = evaluate_ppl_of_prefill_decode_model_on_dataset(q_m, q_m, calib_inputs, pad_token_id=tokenizer.pad_token_id, max_seq_len = args.max_seq_len, seed = args.seed, device = device)
+    #
+    #        print("\n┌── Wikitext-2 prefill_decode initial calibration perplexity──")
+    #        print(f"│ FP32 : {prefill_decode_ppl:8.2f}")
+    #        print("└───────────────────────────────────────────")
+
+    def run_ptq(self, q_m, calib_inputs):
```
Comment (mhs4670go) [inline tico/quantization/wrapq/examples/quantize_full_qmodel_with_gptq.py]: I think the decode calibration data should ideally be generated from the float/reference prefill path, not from the already-quantized prefill model.

If decode calibration inputs are produced by the quantized prefill model, then the KV cache and related decode inputs already contain prefill quantization error. In that case, decode PTQ is calibrated on a distorted distribution rather than on the original float decode distribution.

A cleaner approach would be:

1. run prefill calibration batches through the float/reference model,
2. collect float decode inputs (especially past_key_values),
3. use those inputs for decode PTQ calibration.

That way, decode calibration remains aligned with the original model behavior, and prefill quantization error does not leak into the decode calibration set.

Comment (mhs4670go) [reply tico/quantization/wrapq/examples/quantize_full_qmodel_with_gptq.py]: The reason to prefer float-based prefill for decode calibration is that PTQ fundamentally tries to approximate the float model behavior. If decode calibration inputs are generated from a quantized prefill model, the input distribution is already distorted, and the optimization target effectively changes. This mixes prefill quantization error into decode calibration and makes it harder to isolate and properly optimize decode quantization.

That being said, it would be useful to compare both approaches, so I’m opening [this issue](https://github.com/Samsung/TICO/issues/622) to track it.

File: tico/quantization/wrapq/examples/quantize_full_qmodel_with_gptq.py [hunk 2]
```diff
@@ -94,14 +105,348 @@ def inject_gptq_qparams(
...
+    # fill in input
+    inputs = {}
+    inputs["input_ids"] = torch.tensor([[next_token]])
+    inputs["attention_mask"] = attention_mask
+    inputs["position_embeddings"] = position_embeddings
+    inputs["past_key_values"] = outputs.past_key_values
+    return inputs
+
+
+def evaluate_ppl_of_model_on_dataset(model, dataset, device: str = "cuda"):
```
Comment (mhs4670go) [inline tico/quantization/wrapq/examples/quantize_full_qmodel_with_gptq.py]: IMHO, I think the current evaluation API should be simplified.

Right now, we have multiple similarly named functions such as:

* `evaluate_ppl_of_model_on_dataset`
* `evaluate_ppl_of_ref_prefill_model_on_dataset`
* `evaluate_ppl_of_prefill_ref_model_on_dataset`
* `evaluate_ppl_of_prefill_decode_model_on_dataset`

This makes the API hard to follow because the function names encode multiple dimensions at once:

* metric (`ppl`)
* model role (`ref` / `quant`)
* execution path (`prefill` / `prefill_decode`)

A cleaner design would be to expose a single public API such as:

```python
evaluate_ppl(
    model,
    dataset,
    *,
    model_role="ref" | "quant",
    execution_mode="prefill" | "prefill_decode",
)
```

If it's hard to refactor them once, I'll do this later when the codes are merged.

File: tico/quantization/wrapq/examples/quantize_full_qmodel_with_gptq.py (hunk 1 above)
Comment (stamalakhov) [reply tico/quantization/wrapq/examples/quantize_full_qmodel_with_gptq.py]: Due to #618, this PR should be reworked. A far as i understand, we don't need two torch models. Just single prefill-decode torch model, calibrated with `use_cache` enabled . So i'll be back wirh reworked PR.

File: tico/quantization/wrapq/examples/quantize_full_qmodel_with_gptq.py (hunk 2 above)
Comment (stamalakhov) [reply tico/quantization/wrapq/examples/quantize_full_qmodel_with_gptq.py]: @mhs4670go Sorry for a lack of details. We need just `evaluate_ppl_of_prefill_decode_model_on_dataset`, `evaluate_ppl_of_ref_prefill_model_on_dataset/evaluate_ppl_of_prefill_ref_model_on_dataset ` will be removed. Thank you for your review.
